devdocs/schema_guidelines.md

```diff
@@ -0,0 +1,105 @@
+
+## When to Consider Custom Types
+
+Custom attribute types can add complexity, so try to use built-in Terraform types with validators first.
+
+Reserve custom types for complex scenarios requiring special parsing or logic. For instance, the provider uses a custom UUID type (`customtypes.UUID`) specifically for handling GUID/UUID values, ensuring proper parsing and validation of identifiers across resources and data sources.
+
+## Effectively Using Plan Modifiers
+
+Plan modifiers fine-tune Terraform’s planned changes before they're applied:
+
+- **`UseStateForUnknown`:** Commonly used for computed fields like resource IDs. For example, the `id` field in `powerplatform_data_record` uses this modifier to preserve existing state values. Don't use this modifier when the attribute's value might genuinely need to change or when the API explicitly resets or regenerates the value, as this could mask necessary updates.
+- **`RequiresReplace`:** Clearly indicates changes that necessitate recreating the resource, such as altering `environment_id` or `table_logical_name`.
+- **Dynamic Defaults:** Explicitly set predictable defaults using plan modifiers to avoid confusion in Terraform plans.
+
+Proper use ensures Terraform plans remain predictable and user-friendly.
+
+## Implementing Validators
+
+Validators ensure inputs meet constraints before Terraform applies changes. Clearly differentiate validator use based on your scenario:
+
+- **Built-in Validators:** Use these for straightforward constraints like string length, regex patterns, or predefined enumerations. For instance, attributes like `environment_type` are suitable for built-in validation since new options usually necessitate broader schema updates.
+- **Resource-level Validators:** Employ these when validation logic involves multiple interdependent attributes, such as mutually exclusive flags (`allow_bing_search` vs. `allow_moving_data_across_regions`).
+- **Custom Validators:** Implement custom validators only for complex validation scenarios not covered by built-in options, ensuring validations remain efficient and maintainable.
+
+Balance strictness and flexibility by validating only what you know to be invalid. Avoid overly constraining attributes to allow flexibility for backend changes that might introduce new valid values without requiring schema updates.
+
+## Structuring Schemas with Nested Attributes (Blocks)
+
+Nested blocks logically group related attributes, improving clarity:
+
+- **SingleNested Blocks:** Used for singular configurations, such as Dataverse settings within the `powerplatform_environment` resource, grouping attributes like `language_code`, `currency_code`, and `security_group_id`.
+- **SetNested or ListNested Blocks:** Used for repeatable structures, such as multiple connector references or multiple similar configurations.
+
+Avoid overly deep nesting. Clearly represent resource or data source structure without unnecessary complexity.
```

Comment:
I would add more info about judgment, how to structure a new resource. Right now if something has a complex structure like for example env settings, we try to follow the ppac UI in our schema.